### ISSUE TITLE
How long between typical write <-> read commands over SSH?

### ISSUE THREAD

**Issue by cryptogenesis** [phpseclib/phpseclib#1080]
I'm seeing about 8 seconds of "idle" in-between when various commands are written to the channel over SSH -- is this standard or am I missing something configuration-wise?

Thank you in advance!

**Comment by terrafrost** [comment]
What are the commands you're trying to run? Some commands just take a long time to run. Example:

```php
$ssh->write("sleep 5\n");
```
(idk if you're using the interactive methods or `exec()`)

I suppose you could enable logging and post the log and more info could be provided. You can do that by doing `define('NET_SSH2_LOGGING', 2);` at the top of the file and then `echo $ssh->getLog();` after it's done everything it needs to do.

That said, here's another possibility (in lieu of knowing anything about what you're trying to do): the SSH channels are just super noisy. Example:

```php
$ssh->write("cat /dev/urandom\n");

$ssh->setTimeout(10);
$start = microtime(true);
echo $ssh->exec('ping google.com');
$elapsed = microtime(true) - $start;

echo "took $elapsed seconds\n";
```
That takes 20s despite the timeout because it's spending a lot of time decoding the junk it's getting on another channel.

**Comment by cryptogenesis** [comment]
Thank you for the reply! I'm using standard $ssh->write and then $ssh->read, I enabled logging and here's what some of the output looks like (I imagine the since last: seems to show this delay I am experiencing?)

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.4278, network: 0.0001s)
00000000  00:00:00:01:00:00:00:13:75:62:75:6e:74:75:40:6c  ........ubuntu@l
00000010  69:32:30:34:2d:36:35:3a:7e:24:20                 i204-65:~$ 

-> NET_SSH2_MSG_CHANNEL_DATA (since last: 9.1794, network: 0.0001s)
00000000  00:00:00:00:00:00:00:07:6c:73:20:2d:6c:61:0a     ........ls -la.

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:13:75:62:75:6e:74:75:40:6c  ........ubuntu@l
00000010  69:32:30:34:2d:36:35:3a:7e:24:20                 i204-65:~$ 

-> NET_SSH2_MSG_CHANNEL_DATA (since last: 10.0057, network: 0.0001s)
00000000  00:00:00:00:00:00:00:08:73:75:64:6f:20:73:75:0a  ........sudo su.

**Comment by terrafrost** [comment]
It's still hard to say for sure without seeing your code.

If you're doing `$ssh->read()`, without actually passing any parameters to `read()` then yah - it'll only return when it times out.

The thing is...  phpseclib doesn't know if what it's opening is a shell or not. It doesn't know what the prompt looks like if even there is one. All phpseclib knows is that data is coming at it and could come at it at any time.

Here are a few scenarios that better illustrate this.

1. You've SSH'd into a box and your console is just sitting on the prompt. You're not doing anything. Then someone shuts the system down. At that point everyone will get a message, regardless of whether or not they're actually doing something at that moment in time, that the system is shutting down. ie. the server can send data even if the client doesn't ask for it. And phpseclib doesn't have any way of knowing that you /don't/ want to wait 10s for (for example) shutdown broadcast messages.

2. Let's say your .bashrc has vim in it and that's it. Then when you SSH into a machine and then do `$ssh->write('...')` it's not bash that's interpreting the commands - it's vim. And similarly, there isn't really a prompt, per say. phpseclib doesn't know if, when you SSH in, if you're in a real shell or if you're in a some sort of app like vim. phpseclib doesn't know what type of response the various bytes of data you send over should do.

So, long story short, if you're doing `$ssh->read()`, without parameters, try `$ssh->read('ubuntu@1~$');` or whatever it is that your prompt is, instead.
